server/storage/wal/wal.go

```diff
@@ -396,6 +396,65 @@ func openAtIndex(lg *zap.Logger, dirpath string, snap walpb.Snapshot, write bool
 	return w, nil
 }
 
+func ReadAllEntries(lg *zap.Logger, dirpath string) (state raftpb.HardState, ents []raftpb.Entry, err error) {
+	names, nameIndex, err := selectWALFiles(lg, dirpath, walpb.Snapshot{Term: 0, Index: 0})
+	if err != nil {
+		return state, nil, fmt.Errorf("[ReadAllEntries] selectWALFiles failed: %w", err)
+	}
+
+	rs, _, closer, err := openWALFiles(lg, dirpath, names, nameIndex, false)
+	if err != nil {
+		return state, nil, fmt.Errorf("[ReadAllEntries] openWALFiles failed: %w", err)
+	}
+	defer closer()
+	rec := &walpb.Record{}
+	decoder := NewDecoder(rs...)
+	var startIndex uint64
+	for err = decoder.Decode(rec); err == nil; err = decoder.Decode(rec) {
+		switch rec.Type {
+		case EntryType:
+			e := MustUnmarshalEntry(rec.Data)
+			if e.Index == 0 {
+				return state, nil, fmt.Errorf("entry with index 0 found")
+			}
+			if startIndex == 0 {
+				startIndex = e.Index
+				ents = append(ents, e)
+				continue
+			}
+			if e.Index > startIndex {
+				offset := e.Index - startIndex
+				if offset > uint64(len(ents)) {
+					ents = append(ents, make([]raftpb.Entry, offset-uint64(len(ents)))...)
+				}
+				// The line below is potentially overriding some 'uncommitted' entries.
+				ents = append(ents[:offset], e)
+			}
+		case StateType:
+			state = MustUnmarshalState(rec.Data)
+		case MetadataType:
```

Comment:
I think we don't need these cases.

Reply:
I'm not sure why GitHub is highlighting the wrong lines...  I meant the blank cases like MetadataType, CrcType, and SnapshotType.

Reply:
Added handling of CRC because it would not be great if we got a flakes because of corruption and not be aware of it.